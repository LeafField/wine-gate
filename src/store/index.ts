import { create } from "zustand";

type User = {
  username: string;
  id: string;
  email: string | undefined;
};

type Menu = "close" | "member" | "free" | "category";

type Store = {
  user: User | null;
  setUser: (user: User) => void;
  logout: () => void;
  menu: Menu;
  setMenu: (menu: Menu) => void;
};

export const useStore = create<Store>((set) => ({
  user: null,
  setUser: (user) => set({ user }),
  logout: () => set({ user: null }),
  menu: "close",
  setMenu: (menu) => set({ menu }),
}));
