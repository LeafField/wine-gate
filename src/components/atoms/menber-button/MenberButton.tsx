"use client";
import React, { FC } from "react";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faUser } from "@fortawesome/free-regular-svg-icons";
import { useStore } from "../../../store";
import CloseButton from "../close-button/CloseButton";

const MenberButton: FC = () => {
  const { menu, setMenu } = useStore();

  const handleClick = () => {
    setMenu("member");
  };

  if (menu === "member") {
    return <CloseButton rightBorder={true} />;
  } else {
    return (
      <button
        type="button"
        className="flex w-full flex-col items-center justify-center border-r border-white"
        onClick={handleClick}
        aria-controls="user-panel"
        aria-expanded={false}
      >
        <FontAwesomeIcon
          className="text-menu-icon text-white"
          icon={faUser}
          aria-hidden={"true"}
        />
        <span className="text-menu-title text-white">会員メニュー</span>
      </button>
    );
  }
};

export default MenberButton;
