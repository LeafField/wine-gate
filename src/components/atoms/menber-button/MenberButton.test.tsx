import {
  screen,
  render,
  waitFor,
  renderHook,
  act,
} from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import MenberButton from "./MenberButton";
import { useStore } from "../../../store";

describe("MenberButtonの結合テスト", () => {
  beforeAll(() => {
    const { result } = renderHook(() => useStore());
    act(() => {
      result.current.setMenu("close");
    });
  });

  test("初期状態では会員メニューのボタンが表示されている", () => {
    const { container } = render(<MenberButton />);
    expect(screen.getByRole("button")).toHaveTextContent("会員メニュー");
    expect(container).toMatchSnapshot();
  });

  test("会員メニューのボタンをクリックすると、会員メニューのボタンが閉じるボタンに変わる", async () => {
    render(<MenberButton />);
    await userEvent.click(screen.getByRole("button"));
    return expect(screen.getByRole("button")).toHaveTextContent("閉じる");
  });

  test("閉じるボタンをクリックすると、会員メニューのボタンに変わる", async () => {
    render(<MenberButton />);
    await userEvent.click(screen.getByRole("button", { name: "閉じる" }));
    await waitFor(() => {
      expect(screen.getByRole("button")).toHaveTextContent("会員メニュー");
    });
  });
});
