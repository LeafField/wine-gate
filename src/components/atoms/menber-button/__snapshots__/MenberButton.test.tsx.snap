// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MenberButtonの結合テスト 初期状態では会員メニューのボタンが表示されている 1`] = `
<div>
  <button
    class="flex w-full flex-col items-center justify-center border-r border-white"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="svg-inline--fa fa-user text-menu-icon text-white"
      data-icon="user"
      data-prefix="far"
      focusable="false"
      role="img"
      viewBox="0 0 448 512"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M304 128a80 80 0 1 0 -160 0 80 80 0 1 0 160 0zM96 128a128 128 0 1 1 256 0A128 128 0 1 1 96 128zM49.3 464H398.7c-8.9-63.3-63.3-112-129-112H178.3c-65.7 0-120.1 48.7-129 112zM0 482.3C0 383.8 79.8 304 178.3 304h91.4C368.2 304 448 383.8 448 482.3c0 16.4-13.3 29.7-29.7 29.7H29.7C13.3 512 0 498.7 0 482.3z"
        fill="currentColor"
      />
    </svg>
    <span
      class="text-menu-title text-white"
    >
      会員メニュー
    </span>
  </button>
</div>
`;
